### DISCUSSION TITLE
[Idea] Controller before/after action methods

### DISCUSSION THREAD

**Discussion by aurimasniekis** [symfony/symfony#42764] [Q&A] [answered]
I am not sure if this is the correct place to discuss this but I thought it's a good starting point to discuss the idea I was having for the last couple of months while working on one project.

So the general idea about this is the introduction of `beforeAction` and `afterAction` methods into controllers which would be called before calling action method and after action method if needed. I am aware of the ability to use events for this but I think sometimes u have methods that have almost the same beginning of the method, which could be extracted to a separate method.

Example:

```php
<?php

#[Route("/api/client/{clientId}/items", name: "api.client.items.")]
class ClientItemController extends AbstractController
{
    private Client $client;

    public function __construct(private ClientRepository $clientRepository, private ItemRepository $itemRepository)
    {
    }
    
    #[BeforeAction]
    public function beforeAction(#[RouteParam] string $clientId): void
    {
        // Some common logic of controller class like loading common resources, check required roles
        
        $this->client = $this->clientRepository->find($clientId);
        
        if (null === $this->client) {
            throw new NotFoundHttpException();
        }
        
        $user = $this->getUser();
        
        if (false === $this->client->hasAccess($user)) {
            throw new AccessDeniedHttpException();
        }
    }
    
    #[Route("/", name: "index", methods: ["GET"])]
    public function index(): Response
    {
        // Some action specific logic
    }
    
    #[Route("/{id}", name: "index", methods: ["GET"])]
    public function show(string $id): Response
    {
        $item = $this->itemRepository->findOneBy(['id' => $id, 'client' => $this->client]);

        // Some action specific logic
    }
}
```

This is just my example of the proposal and concrete implementation could be up for discussion. 

I think this would be a great addition to the Symfony framework.

**Comment by chyshyya** [comment]
It's useless.
Your idea must be implemented with creating `BaseController extends AbstractController` that will execute `BeforeAction`. And this ClientItemController will extends BaseController. Thats it.
Simple "Decorator Pattern"

**Reply by aurimasniekis** [reply]
Right, so u would need to wrap each action to execute it? What's then different from just calling it in each method?

**Reply by chyshyya** [reply]
As an example...
Client class just to emulate:
```
<?php

namespace App\Controller;

class Client
{
    private int $id;
    private string $name;

    public function __construct(int $id, string $name)
    {
        $this->id = $id;
        $this->name = $name;
    }

    /**
     * @return int
     */
    public function getId(): int
    {
        return $this->id;
    }

    /**
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @param string $name
     */
    public function setName(string $name): void
    {
        $this->name = $name;
    }
}
```
BaseController class for your specific jobs:
```
<?php
namespace App\Controller;

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\RequestStack;

class BaseItemController extends AbstractController
{
    private array $clients;
    private ?Client $client;

    public function __construct(RequestStack $requestStack)
    {
        $this->clients[1] = new Client(1, 'Name 1');
        $this->clients[2] = new Client(2, 'Name 2');
        $this->clients[3] = new Client(3, 'Name 3');



        $request = $requestStack->getCurrentRequest();

        // Executes before any action
        $this->beforeAlways();

        // Execute if request has clientId attribute in @Route
        if ($request->attributes->has('clientId')) {
            $this->beforeIfClientId($request->attributes->get('clientId'));
        }
    }

    private function beforeAlways(): void
    {
        // Do something
    }

    private function beforeIfClientId(int $clientId): void
    {
        // Do something if clientId presented in @Route attributes
        $this->client = array_key_exists($clientId, $this->clients) ? $this->clients[$clientId] : null;
    }

    protected function getClients(): array
    {
        return $this->clients;
    }

    protected function getClient(): ?Client
    {
        return $this->client;
    }
}
```
And you will have absolutly clear and undestandable code of your controller that do what you want. Simple and clear for other developers :
```
<?php
namespace App\Controller;

use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\Routing\Annotation\Route;

/**
 * @Route("/test")
 */
class ItemController extends BaseItemController
{
    /**
     * @Route("/clients", name="test_clients", methods={"GET"})
     *
     * @return JsonResponse
     */
    public function clients(): JsonResponse
    {
        return $this->json($this->getClients());
    }

    /**
     * @Route("/client/{clientId<\d+>}", name="test_client", methods={"GET"})
     *
     * @return JsonResponse
     */
    public function client(): JsonResponse
    {
        return $this->json($this->getClient());
    }
}
```
Create three classes in `/src/Controller` directory and test it. On my localhost code is working.
That's it.

**Reply by aurimasniekis** [reply]
I think you didn't really understand my post here. I don't need help to make this work, I am proposing a feature addition that allows having a pre-action method in controller classes. I already have some code done as an external library, and I would like to bring this feature into the core. 

P.S. It's bad practice to place logic inside the constructor as classes are initialized inside the container...

**Reply by faizanakram99** [reply]
There is a php lib which does this, AOP (aspect oriented programming), don't remember the name though

**Comment by chuckadams** [comment]
It really does look like a Decorator thing, which is built-in to Symfony, e.g.:

```yaml
# config/services.yaml
services:
    App\Controller\WrapperController:
        decorates: App\Controller\OriginalController
```

I'll admit isn't quite as convenient as the methods, but it would give you even more flexibility in that it can wrap everything about the controller, not just one method.  You could also possibly do it with a trait and a reusable event listener, even turn that into a bundle.  But if I needed general-purpose method advice, I'd probably reach for one of those AOP packages (and hope it's maintained).

**Reply by aurimasniekis** [reply]
There are multiple ways you can achieve this with current tools on hand, but all of them fail with for e.g. having a simple method with a predefined name or just attribute which specifies that this is before/after.

The key point is to keep everything relating in the same spot, not in different classes or multiple configurations.

**Comment by javiereguiluz** [comment, marked as answer]
In Symfony 1 there was a builtin feature to run code before/after controller methods. This was removed, but in modern Symfony you can achieve that with this: https://symfony.com/doc/current/event_dispatcher/before_after_filters.html

**Comment by pedroresende** [comment]
Why not use events? https://symfony.com/doc/current/event_dispatcher.html

**Reply by pedroresende** [reply]
Or before after filter ? https://symfony.com/doc/current/event_dispatcher/before_after_filters.html
